python/whylogs/api/logger/multi_dataset_logger/multi_dataset_rolling_logger.py

```diff
@@ -0,0 +1,362 @@
+from concurrent.futures import Future
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Any, Dict, List, Optional, Union
+
+import pandas as pd
+from dateutil import tz
+
+from whylogs.api.logger.result_set import ProfileResultSet, ResultSet
+from whylogs.api.logger.segment_cache import SegmentCache
+from whylogs.api.logger.segment_processing import segment_processing
+from whylogs.api.store import ProfileStore
+from whylogs.api.writer import Writer, Writers
+from whylogs.api.writer.writer import Writable
+from whylogs.core import DatasetProfile, DatasetSchema
+
+from .future_util import wait_result
+from .message_processor import CloseMessage, MessageProcessor
+from .time_util import (
+    FunctionTimer,
+    Schedule,
+    TimeGranularity,
+    current_time_ms,
+    truncate_time_ms,
+)
+
+Row = Dict[str, Any]
+TrackData = Union[pd.DataFrame, Row, List[Row]]
+
+
+class DatasetProfileContainer:
+    """
+    A container that abstracts over different types of profiles.
+
+    This does the work of deciding how to track data and how to create profiles given a DatasetSchema.
+    This can only be used to manage a single entity for a given time. For example, this can represent
+    a normal DatasetProfile or segment that has a given dataset timestamp.
+    """
+
+    _target: Union[DatasetProfile, SegmentCache]
+
+    def __init__(self, dataset_timestamp: int, schema: Optional[DatasetSchema]) -> None:
+        self._schema: Optional[DatasetSchema] = schema
+        self._active = True
+        self._dataset_timestamp = datetime.fromtimestamp(dataset_timestamp / 1000.0, tz=tz.tzutc())
+        if self._has_segments() and schema is not None:  # Need the duplicate None check for type safety
+            self._target = SegmentCache(schema=schema)
+        else:
+            self._target = DatasetProfile(dataset_timestamp=self._dataset_timestamp)
+
+    def _has_segments(self) -> bool:
+        return self._schema is not None and self._schema.segments is not None
+
+    def _track_segments(self, data: TrackData) -> None:
+        if self._schema is None:
+            raise Exception("Schema missing in logger while using segments")
+
+        if not isinstance(self._target, SegmentCache):
+            raise Exception("Segment cache missing in logger while using segments")
+
+        if isinstance(data, List):
+            for row in data:
+                segment_processing(self._schema, row=row, segment_cache=self._target)
+        else:
+            segment_processing(self._schema, data, segment_cache=self._target)
+
+    def _track_profile(self, data: TrackData) -> None:
+        if not isinstance(self._target, DatasetProfile):
+            raise Exception("Dataset profile missing in logger")
+
+        if isinstance(data, List):
```

Comment:
what exactly do you expect here? a List of dicts or so?

Reply:
No its duplicate type checking. You need this for type safety even though it can't be anything else at this point. Basically sanity checks that should never happen because we catch them in unit tests if we violate those constriants.